First pass at adapter registration support

Blacksmoke16 opened the pull request:
First pass on #254.  Tests pass so assuming this works.  Still some things to do tho :p Feedback appreciated :)

TODOs:
* Write tests for this
* Possibly cleanup `Granite.adapters.adapters` syntax
* Update docs

Blacksmoke16 commented:
@drujensen @robacarp How do you think I should best test this?  In theory I shouldn't have to actually test connecting to diff databases as that already happens for each adapter macro.  

Probably just add some connections and ensure that it finds right one?

drujensen commented:
Maybe we can add multiple database testing #249 with this effort and leverage different adapters to the same db type.  wdyt?

drujensen commented on src/granite/settings.cr:
```diff
@@ -2,16 +2,15 @@ require "logger"
-    property database_url : String? = nil
-    property logger : Logger
+    class_getter adapters = [] of Granite::Adapter::Base
+    class_property logger = Logger.new STDOUT
```
I wonder if we should change this to `::Logger.new(nil)`.  This potentially could be a bottleneck if all logging is sent to a single IO class instance.  @robacarp wdyt?

Blacksmoke16 replied:
Does having it at the class level matter?  Before, `@logger` was defaulted to `Logger.new STDOUT` as well.  Was just keeping that behavior.

robacarp replied:
In practice this will almost always be initialized to STDOUT, or a logger pointed at STDOUT. But I think the default should be nil. It's frustrating when including a library suddenly generates a bunch of output in a program and requires additional configuration to silence it. For example, imagine including Granite in a console application which reads sqlite files -- Granite would pollute the user interface by default. The logger facility is a great tool for use of granite, but silence feels like a nice default.

drujensen commented on src/granite/table.cr:
```diff
@@ -9,11 +9,7 @@ module Granite::Table
-    def self.adapter
-      @@adapter
-    end
+    class_getter adapter : Granite::Adapter::Base = Granite::Settings.adapters.find { |a| a.name == {{name.stringify}} } || raise "No registered adapter with the name '{{name.id}}'"
```
nice!

Blacksmoke16 commented:
@drujensen That might be the best option IMO.  The more I think about this, i'm thinking that it gets tested by default.  Otherwise none of the 3 adapters we provide would connect and all the tests would fail.

Testing on a different versions would be a good addition.  Deff work pursuing.

Blacksmoke16 commented:
Alright @drujensen @robacarp @eliasjpr This is ready for review now.

robacarp commented on spec/adapter/base_spec.cr:
```diff
@@ -1,32 +1,40 @@
+        sqlite_adapter.url.should eq ENV["SQLITE_DATABASE_URL"]
+      end
+      if foo_adapter = Granite::Settings.adapters.find { |adapter| adapter.name == "FooBar123" }
+        foo_adapter.url.should eq "FAKE_DB"
```
these tests can fail with a false positive if the find returns nil. Is there an easy way to write them so that they'll throw when find returns nil?

Blacksmoke16 replied:
Could just do the find off adapters array and `not_nil!` it?

robacarp commented on spec/adapter/base_spec.cr:
```diff
@@ -1,32 +1,40 @@
-        ENV["DATABASE"] = "test"
-        url = "dummy://user:pswd@host:3333/${DATABASE}"
-        Granite::Adapter::Base.env(url).should eq "dummy://user:pswd@host:3333/test"
+    it "should disallow adapters multiple adapters with the same name" do
```
should this be "should disallow multiple adapters with the same name"?

robacarp commented on spec/adapter/base_spec.cr:
```diff
@@ -1,32 +1,40 @@
-        url = "dummy://user:pswd@host:3333/database"
-        Granite::Adapter::Base.env(url).should eq url
+  describe "registration" do
+    it "should allow adapters to be registered" do
```
it "should allow adapters to be saved and looked up"

robacarp commented on spec/adapter/base_spec.cr:
```diff
@@ -1,32 +1,40 @@
-      it "should return itself" do
-        url = "dummy://user:pswd@host:3333/database"
-        Granite::Adapter::Base.env(url).should eq url
+  describe "registration" do
```
Since the tests in this file mess with the global scope of tests, it has some potential to create problems with other tests when run in different orders. Is it possible to use some test interface to run these tests without depending on the global scope of Granite::Settings?

Blacksmoke16 replied:
Possibly?  Really i could probably refactor that to use a one of the predefined classes and just test it based off of adapters already registered.  

i.e.

```Crystal
class Foo
  adapter mysql
end

adapter = Foo.adapter
adapter.name.should eq "mysq"
adapter.url.should eq ENV['MYSQL_DATABASE_URL"]
```

Would that be better?

robacarp commented on src/adapter/base.cr:
```diff
@@ -5,34 +5,21 @@ require "db"
-      return {"database": ""}
-    end
+  def initialize(connection_hash : NamedTuple(url: String, name: String))
+    raise "Adapter with name '#{connection_hash[:name]}' has already been registered." if Granite::Settings.adapters.any? { |adapter| adapter.name == connection_hash[:name] }
```
This check logic feels out of place here, because it feels like this specific instance of an adapter shouldn't have knowledge of the rest of the adapters. Is there another place it could live where it's not crossing domain boundaries?

Blacksmoke16 replied:
Not sure where else to put it.  Can't override the `<<` method in the Settings class so this was only other place that kinda worked since it has access to name of adapter being instantiated.

robacarp commented on src/granite/settings.cr:
```diff
@@ -2,16 +2,11 @@ require "logger"
   class Settings
-    property database_url : String? = nil
-    property logger : Logger
+    class_getter adapters = [] of Granite::Adapter::Base
```
Why should these be changed from properties of a class to class properties? This feels like it will set Granite back a few steps in composability.

Blacksmoke16 replied:
I'm not sure which is the preferred way of doing this.  I assumed ```@@settings ||= Settings.new` was kind of a hack and this was the better method?

If not i can change it back if you'd like.

Blacksmoke16 commented:
Refactored a bit:

* Syntax is now `Granite::Adapters << adapter`
  * The `Adapters' <<` class method now hold the duplicate name logic
* Changed the settings back to how it was before
  * Defaulted logger to nil

Blacksmoke16 commented:
For how easy it would be to do might not be a bad idea.

drujensen commented:
@Blacksmoke16 this is excellent work.  Yes, I am for making the models as simple as possible. I think it would need to raise an exception if there are 2 or more adapters and no adapter is specified, right?

Blacksmoke16 commented:
More i'm looking into this, the main issue is we are using the `adapter` macro to set the adapter for a given model.  But if that is omitted what would set the adapter?

robacarp commented:
@blacksmoke16 hmm good question. Maybe we could implement the adapter method on the base class to find the default adapter and raise if there’s an ambiguity, and then have the adapter macro define a new method which explicitly retrieves the specified adapter.

drujensen commented:
Let’s do this in a separate PR. 

> On Jul 22, 2018, at 8:33 AM, robacarp <notifications@github.com> wrote:
> 
> @Blacksmoke16 hmm good question. Maybe we could implement the adapter method on the base class to find the default adapter and raise if there’s an ambiguity, and then have the adapter macro define a new method which explicitly retrieves the specified adapter.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
